homeassistant/components/thermostat/ecobee.py

```diff
@@ -0,0 +1,263 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+ECOBEE_CONFIG_FILE = 'ecobee.conf'
+_CONFIGURING = {}
+
+
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """ Setup Platform """
+    # Only act if we are not already configuring this host
+    if 'ecobee' in _CONFIGURING:
+        return
+
+    setup_ecobee(hass, config, add_devices_callback)
+
+
+def setup_ecobee(hass, config, add_devices_callback):
+    """ Setup ecobee thermostat """
+    from pyecobee import Ecobee, config_from_file
+    # Create ecobee.conf if it doesn't exist
+    if not os.path.isfile(hass.config.path(ECOBEE_CONFIG_FILE)):
+        jsonconfig = {"API_KEY": config[CONF_API_KEY]}
+        config_from_file(hass.config.path(ECOBEE_CONFIG_FILE), jsonconfig)
+
+    ecobee = Ecobee(hass.config.path(ECOBEE_CONFIG_FILE))
+
+    # If ecobee has a PIN then it needs to be configured.
+    if ecobee.pin is not None:
+        # ecobee.request_pin()
+        request_configuration(ecobee, hass, add_devices_callback)
+        return
+
+    if 'ecobee' in _CONFIGURING:
+        _CONFIGURING.pop('ecobee')
+        configurator = get_component('configurator')
+        configurator.request_done('ecobee')
+
+    devices = []
+    for index in range(0, len(ecobee.thermostats)):
+        devices.append(Thermostat(ecobee, index))
+
+    add_devices_callback(devices)
+
+
+def request_configuration(ecobee, hass, add_devices_callback):
+    """ Request configuration steps from the user. """
+    configurator = get_component('configurator')
+    if 'ecobee' in _CONFIGURING:
+        configurator.notify_errors(
+            _CONFIGURING['ecobee'], "Failed to register, please try again.")
+
+        return
+
+    # pylint: disable=unused-argument
+    def ecobee_configuration_callback(data):
+        """ Actions to do when our configuration callback is called. """
+        ecobee.request_tokens()
+        ecobee.update()
+        setup_ecobee(hass, None, add_devices_callback)
+
+    _CONFIGURING['ecobee'] = configurator.request_config(
+        hass, "Ecobee", ecobee_configuration_callback,
+        description=(
+            'Please authorize this app at https://www.ecobee.com/consumer'
+            'portal/index.html with pin code: ' + ecobee.pin),
+        description_image='https://goo.gl/6tBkbH',
+        submit_caption="I have authorized the app."
+    )
+
+
+class Thermostat(ThermostatDevice):
+    """docstring for Thermostat"""
+
+    def __init__(self, ecobee, thermostat_index):
+        self.ecobee = ecobee
+        self.thermostat_index = thermostat_index
+        self.thermostat_data = self.ecobee.get_thermostat(
+            self.thermostat_index)
+        self._name = self.thermostat_data['name']
+        if 'away' in self.thermostat_data['program']['currentClimateRef']:
+            self._away = True
+        else:
+            self._away = False
+
+    def update(self):
+        self.thermostat_data = self.ecobee.get_thermostat(
```

Comment:
It's fine to do a full refresh, just not for every thermostat. That's n-1
unnecessary refreshes each update cycle. Using an approach in HA similar to
bitcoin will prevent you from updating too often.

You don't have to be afraid of accessing the config simultaneously because
there is no threading during home Assistant boot.

On Wed, Nov 18, 2015, 07:29 Nolan Gilley <notifications@github.com> wrote:

> In homeassistant/components/thermostat/ecobee.py
> <https://github.com/balloob/home-assistant/pull/635#discussion_r45213154>:
>
> > +class Thermostat(ThermostatDevice):
> > +    """docstring for Thermostat"""
> > +
> > +    def __init__(self, ecobee, thermostat_index):
> > +        self.ecobee = ecobee
> > +        self.thermostat_index = thermostat_index
> > +        self.thermostat_data = self.ecobee.get_thermostat(
> > +            self.thermostat_index)
> > +        self._name = self.thermostat_data['name']
> > +        if 'away' in self.thermostat_data['program']['currentClimateRef']:
> > +            self._away = True
> > +        else:
> > +            self._away = False
> > +
> > +    def update(self):
> > +        self.thermostat_data = self.ecobee.get_thermostat(
>
> @balloob <https://github.com/balloob> First off, thank you so much for
> the code review. It helps a lot.
>
> As for the update() method being expensive since it does a full API
> refresh, how would you recommend I do it? I know ecobee does have an API
> call you can make to see if anything has changed since the last request. I
> could do that and only do the full refresh if things have changed.
>
> As for the sensors, I am going to work on cleaning up python-ecobee to
> make this work better. I had originally planned on the sensor using it's
> own ecobee API object, but I was afraid that the 2 objects might clash
> while writing/reading tokens from the config file.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/635/files#r45213154>.
>